File: Data Structures in C/Linked Lists/Circular Linked List/CircularLinkedListInsertion.c
/*
   This program shows how to insert a node at the begining and at the end of a Circular Linked List using C
   Author - Rajarshi Sengupta
   Github - https://github.com/rajarshisg
   Date - 19/08/2020 (dd/mm/yyyy)
*/
#include<stdio.h>
#include<stdlib.h>
typedef struct Node
{
    int data;
    struct Node *next;
}Node;
Node *insertBegining(Node *head,int data)
{
    Node *n=(Node *)malloc(sizeof(Node));
    n->data=data;
    n->next=n;
    if(head==NULL)
    return n;
     Node *tail=head;
    while(tail->next!=head)
    {
       tail=tail->next;
    }
     n->next=head;
     tail->next=n;
     head=n; 
   return head;
}
Node *insertEnd(Node *head,int d)
{
    Node *n=(Node *)malloc(sizeof(Node));
    n->data=d;
    n->next=n;
    if(head==NULL)
    return n;
     Node *curr=head;
    while(curr->next!=head)
    {
       curr=curr->next;
    }
    curr->next=n;
    n->next=head;
    return head;
}
void traverse(Node *head)
{
     Node *curr=head;
    do
    {
        printf("%d ",curr->data);
        curr=curr->next;
    } while(curr!=head);
    printf("\n");
}
int main()
{
    int num;
    printf("Enter the number of nodes :\n");
    scanf("%d",&num);
    Node *head=NULL;
    Node *n=NULL;
    Node *m=NULL;
    printf("Enter the elements :\n");
    for(int i=1;i<=num;i++)
    {
        if(head==NULL)
        {
            n=(Node *)malloc(sizeof(Node));
            scanf("%d",&n->data);
            n->next=NULL;
            head=n;
        }
        else
        {
            m=(Node *)malloc(sizeof(Node));
            scanf("%d",&m->data);
            m->next=NULL;
            n->next=m;
            n=m;
        }
        if(i==num)
        n->next=head;
    }
    printf("The Linked List is :\n");
    traverse(head);
    printf("Enter the data you want to insert at begining :\n");
    int data;
    scanf("%d",&data);
    head=insertBegining(head,data);
    traverse(head);
    printf("Enter the data you want to insert at end :\n");
    int d;
    scanf("%d",&d);
    head=insertEnd(head,d);
    traverse(head);
    return 0;
}

File: Data Structures in JAVA/Linked Lists/Circular Linked Lists/src/CircularLinkedListInsertion.java
/*
   This program shows how insert a node at the begining and end in a Circular Linked List in JAVA
   Author - Rajarshi Sengupta
   Github - https://github.com/rajarshisg
   Date - 20/08/2020 (dd/mm/yyyy)
   
   
   Node class definition :
   public class Node<T> {
   T data;
   Node next;
   Node(T data){
	   this.data=data;
	   this.next=null;
   }
   
}

*/
import java.util.Scanner;
public class CircularLinkedListInsertion {
	Node insertBegining(Node<Integer> head,int data) {
		Node<Integer> n=new Node<Integer>(data);
		Node<Integer> tail=head;
	    while(tail.next!=head)
	    	tail=tail.next;
		n.next=head;
		tail.next=n;
		head=n;
		return head;
	}
	Node insertEnd(Node<Integer> head,int data) {
		Node<Integer> n=new Node<Integer>(data);
		Node<Integer> tail=head;
	    while(tail.next!=head)
	    	tail=tail.next;
		n.next=head;
		tail.next=n;
		return head;
	}
	void traverse(Node<Integer> head) {
		 System.out.println("The Linked List is :");
		   Node curr=head;
		   do
		   {
			   System.out.print(curr.data+" ");
			   curr=curr.next;
		   }while(curr!=head);
		   System.out.println();
	}
   public static void main(String args[])
   {
	   Scanner sc=new Scanner(System.in);
	   Node<Integer> head=null,n=null,m=null;
	   System.out.println("Enter the number of nodes :");
	   int num=sc.nextInt();
	   System.out.println("Enter the elements :");
	   for(int i=1;i<=num;i++)
	   {
		   if(head==null)
		   {
			   n=new Node(sc.nextInt());
			   head=n;
		   }
		   else
		   {
			   m=new Node(sc.nextInt());
			   n.next=m;
			   n=n.next;
		   }
	   }
	   n.next=head; //last node points to first
	   CircularLinkedListInsertion obj=new CircularLinkedListInsertion();
	   obj.traverse(head);
	   System.out.println("Enter the node you want to insert at begining :");
	   int data=sc.nextInt();
	   head=obj.insertBegining(head,data);
	   obj.traverse(head);
	   System.out.println("Enter the node you want to insert at end :");
	   data=sc.nextInt();
	   head=obj.insertEnd(head,data);
	   obj.traverse(head);
   }
}
